### ISSUE TITLE
API request: error code for cancellation

### ISSUE THREAD

**Issue by cheshire** [Z3Prover/z3#431]
Hi,

When using an error handler we need to differentiate between exception arising from cancellations, and exceptions arising from actual errors.
From C API the only way we could do that is by comparing the string returned by Z3_get_error_code to "canceled".
This is a bit scary, since the string "canceled" seems to be hardcoded in quite a few places inside Z3.

Would it be possible to add cancellation to Z3_error_code instead?

**Comment by wintersteiger** [comment]
Sounds like a good idea to me. Alternatively, it might also make sense for us to catch the exception internally and not throw it across the API, so that `cancelled' looks like Z3_OK.

**Comment by cheshire** [comment]
> and not throw it across the API, so that `cancelled' looks like Z3_OK.

How would I catch it then?

I would be OK with comparing the returned string to "canceled" if it wouldn't be hard-coded in many places.
Common error messages are in util/common_msgs.cpp, maybe it would make sense to expose them through API and only refer to the constants from there?

**Comment by wintersteiger** [comment]
Yes, forget the alternative, that would require us to add another function to check for cancellation. However, we do already have `Z3_solver_get_reason_unknown`, could that do the job for you?

Total agreement on the hardcoded strings - I'll change those right away.

**Comment by wintersteiger** [comment]
Bumping this; are you OK with the current solution, or are there more issues with this?

**Comment by delcypher** [comment]
It would be nice if the cancelled string was exposed by the Z3 C API so I could to a ``strcmp()`` with that rather than having to hard code possible responses from Z3.

Here are some examples in KLEE where I have to hard code strings that may be returned by ``z3_get_error_msg()`` and ``Z3_solver_get_reason_unknown()``

https://github.com/klee/klee/blob/master/lib/Solver/Z3Builder.cpp#L29

https://github.com/klee/klee/blob/master/lib/Solver/Z3Solver.cpp#L277

It would be nice if there was a cleaner way to do this.

**Comment by NikolajBjorner** [comment]
see Z3_get_error_code
